openexr-rs/src/core/preview_image.rs

```diff
@@ -1,23 +1,470 @@
+    type Pointee = PreviewImage;
+}
+
+pub type PreviewImageRef<'a, P = PreviewImage> = Ref<'a, P>;
+pub type PreviewImageRefMut<'a, P = PreviewImage> = RefMut<'a, P>;
+
+impl PreviewImage {
+    /// Creates a new `PreviewImage` with dimensions width = 0 and height = 0
+    ///
+    /// ## Errors
+    /// * [`Error::Base`] - If an error occurs
+    ///
+    pub fn empty() -> Result<PreviewImage> {
+        PreviewImage::with_dimensions(0, 0)
+    }
+
+    /// Creates a new `PreviewImage` with the provided dimensions
+    ///
+    /// ## Errors
+    /// * [`Error::Overflow`] - If the size of the image is bigger than the supported size
+    /// * [`Error::Base`] - If an error occurs
+    ///
+    pub fn with_dimensions(width: u32, height: u32) -> Result<PreviewImage> {
+        let mut _inner = std::ptr::null_mut();
+        unsafe {
+            sys::Imf_PreviewImage_ctor(
+                &mut _inner,
+                width,
+                height,
+                std::ptr::null(),
+            )
+            .into_result()?;
+        }
+
+        Ok(PreviewImage(_inner))
+    }
+
+    /// Creates a new `PreviewImage` with the provided dimensions and pixels
+    ///
+    /// ## Errors
+    /// * [`Error::Overflow`] - If the size of the image is bigger than the supported size
+    /// * [`Error::Base`] - If an error occurs
+    ///
+    pub fn with_pixels(
+        width: u32,
+        height: u32,
+        pixels: &[PreviewRgba],
+    ) -> Result<PreviewImage> {
+        let mut _inner = std::ptr::null_mut();
+        unsafe {
+            sys::Imf_PreviewImage_ctor(
+                &mut _inner,
+                width,
+                height,
+                pixels.as_ptr() as *const sys::Imf_PreviewRgba_t,
+            )
+            .into_result()?;
+        }
+
+        Ok(PreviewImage(_inner))
+    }
+
+    /// Returns the `width` of the `PreviewImage`
+    ///
+    pub fn width(&self) -> u32 {
+        let mut value = 0;
+        unsafe {
+            sys::Imf_PreviewImage_width(self.0, &mut value)
+                .into_result()
+                .expect("Error getting 'width' property");
+        }
+        value
+    }
+
+    /// Returns the `height` of the `PreviewImage`
+    ///
+    pub fn height(&self) -> u32 {
+        let mut value = 0;
+
+        unsafe {
+            sys::Imf_PreviewImage_height(self.0, &mut value)
+                .into_result()
+                .expect("Error getting 'height' property");
+        }
+
+        value
+    }
+
+    /// Get the image pixels as a slice
+    ///
+    pub fn pixels<'a>(&'a self) -> &'a [PreviewRgba] {
+        let mut pixels = std::ptr::null();
+
+        unsafe {
+            sys::Imf_PreviewImage_pixels_const(self.0, &mut pixels)
+                .into_result()
+                .expect("Error getting 'height' property");
+
+            let pixel_count = (self.width() * self.height()) as usize;
+
+            std::slice::from_raw_parts(

```

Comment:
Is it conceivable that the pixels pointer might not be valid? I'd at least put a if pixels.is_null() { panic!(...) } to guard against getting a nullptr back from C++

Reply:
Updated